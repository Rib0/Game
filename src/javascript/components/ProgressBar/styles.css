.progressBar {
    width: 500px;
    height: 40px;
    margin: 300px auto;
    background: gainsboro;
    border-radius: 10px;
    overflow: hidden;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.25) inset;
}

.progressInner {
    width: 20%;
    height: 100%;
    background: green;
    transition: width 1s linear;

    &.slow {
        transition-duration: 1s;
    }

    &.medium {
        transition-duration: 0.5s;
    }

    &.fast {
        transition-duration: 0.3s;
    }
}
