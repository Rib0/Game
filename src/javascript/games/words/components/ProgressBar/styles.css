.progressBar {
    width: 500px;
    height: 40px;
    margin: 100px auto;
    border: 2px solid var(--colorOrange);
    border-radius: 10px;
    overflow: hidden;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.25) inset;
}

.progressInner {
    width: 20%;
    height: 100%;
    background: var(--colorOrange);
    transition-property: width;
    transition-timing-function: linear;
}
