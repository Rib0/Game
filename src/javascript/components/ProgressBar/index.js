import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import styles from './styles.css';

class ProgressBar extends PureComponent {
    static transitionDurations = {
        slow: '1000ms',
        medium: '700ms',
        fast: '400ms',
    };

    constructor(props) {
        super(props);
        const { transitionDuration } = props;

        this.state = {
            width: 100,
            speed: ProgressBar.transitionDurations[transitionDuration],
        };
    }

    componentDidMount() {
        const { speed, width } = this.state;

        this.timerId = setInterval(() => {
            this.setState({
                width: width - 20, // start game
            });
        }, parseInt(speed));

        this.progress.addEventListener('transitionend', this.checkEndGame);
    }

    componentWillUnmount() {}

    checkEndGame = () => {
        const { width } = this.state;

        if (width < 0) {
            console.log('Вы проиграли');
            clearInterval(this.timerId);
        }
    };

    render() {
        const { width, speed } = this.state;

        const progressInnerStyles = {
            width: `${width}%`,
            transitionDuration: speed,
        };

        return (
            <div className={styles.progressBar}>
                <div ref={elem => (this.progress = elem)} style={progressInnerStyles} className={styles.progressInner} />
            </div>
        );
    }
}

ProgressBar.propTypes = {
    width: PropTypes.number,
    transitionDuration: PropTypes.oneOf[Object.keys(ProgressBar.transitionDurations)],
};

ProgressBar.defaultProps = {
    width: 0,
    transitionDuration: 'slow',
};

export default ProgressBar;
