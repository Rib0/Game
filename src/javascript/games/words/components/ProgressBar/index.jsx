import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import styles from './styles.css';

class ProgressBar extends PureComponent {
    componentDidMount() {
        const { handleTransitionEnd } = this.props;

        this.progress.addEventListener('transitionend', handleTransitionEnd);
    }

    componentWillUnmount() {
        const { handleTransitionEnd } = this.props;

        this.progress.removeEventListener('transitionend', handleTransitionEnd);
    }

    render() {
        const { innerStyles } = this.props;

        return (
            <div className={styles.progressBar}>
                <div
                    ref={elem => (this.progress = elem)}
                    style={innerStyles}
                    className={styles.progressInner}
                />
            </div>
        );
    }
}

ProgressBar.propTypes = {
    handleTransitionEnd: PropTypes.func,
    innerStyles: PropTypes.objectOf(PropTypes.string),
};

export default ProgressBar;
