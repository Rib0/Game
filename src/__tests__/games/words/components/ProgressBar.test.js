import React from 'react';
import ProgressBar from 'games/words/components/ProgressBar';

const setUp = props => mount(<ProgressBar {...props} />);

describe('Words game - ProgressBar test', () => {
    let Component, instance, componentDidMount, componentWillUnmount;
    const props = {
        width: '10%',
        transitionDuration: '100ms',
    };

    beforeEach(() => {
        Component = setUp(props);
        instance = Component.instance();
        componentDidMount = jest.spyOn(ProgressBar.prototype, 'componentDidMount');
        componentWillUnmount = jest.spyOn(ProgressBar.prototype, 'componentWillUnmount');
    });
    afterEach(() => {
        componentDidMount.mockRestore();
        componentWillUnmount.mockRestore();
    });

    test('should render', () => {
        expect(Component).toMatchSnapshot();
    });
    test('should have ref', () => {
        const ref = instance.progress;
        expect(ref).toBeTruthy();
    });
    test('should contain 2 divs', () => {
        const divs = Component.find('div');
        expect(divs).toHaveLength(2);
    });
    test('AddEventListener should called once', () => {
        const ref = instance.progress;
        const addEventListener = jest.spyOn(ref, 'addEventListener');
        instance.componentDidMount();
        expect(addEventListener).toHaveBeenCalledTimes(1);
    });
    test('RemoveEventListener should called once', () => {
        const ref = instance.progress;
        const removeEventListener = jest.spyOn(ref, 'removeEventListener');
        Component.unmount();
        expect(removeEventListener).toHaveBeenCalledTimes(1);
    });
    test('ComponentDidMount should called once', () => {
        instance.componentDidMount();
        expect(componentDidMount).toHaveBeenCalledTimes(1);
    });
    test('componentWillUnmount should called once', () => {
        Component.unmount();
        expect(componentWillUnmount).toHaveBeenCalledTimes(1);
    });
});
