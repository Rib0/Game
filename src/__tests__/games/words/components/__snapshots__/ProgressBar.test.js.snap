// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Words game - ProgressBar test should render 1`] = `
<ProgressBar
  transitionDuration="100ms"
  width="10%"
>
  <div
    className="progressBar"
  >
    <div
      className="progressInner"
      style={
        Object {
          "transitionDuration": "100ms",
          "width": "10%",
        }
      }
    />
  </div>
</ProgressBar>
`;
